File: CS_dataset/spring/cs-2024-3388/Basik_Valeriya_lb2/src/lab2.py
class Node:
    def __init__(self, data, next=None):
        self.data = data
        self.next = next
        pass

    def get_data(self):
        return self.data
        pass

    def change_data(self, new_data):
        self.data = new_data
        pass

    def __str__(self):
        if self.next != None:
            return f"data: {self.data}, next: {self.next.data}"
        else: return f"data: {self.data}, next: None"
        pass


class LinkedList:
    def __init__(self, head=None):
        self.head = head
        if self.head!=None: self.length = 1
        else: self.length = 0
        pass

    def __len__(self):
        return self.length
        pass

    def append(self, element):
        a = Node(element)
        if self.head!=None:
            cur = self.head
            for i in range(self.length-1):
                cur = cur.next
            cur.next = a
        else: self.head = a
        self.length = self.length + 1
        pass

    def __str__(self):
        if self.length == 0:
            return f"LinkedList[]"
        else:
            st = f"LinkedList[length = {self.length}, [{str(self.head)}"
            cur = self.head
            for i in range(self.length-1):
                cur = cur.next
                st +=';'+ ' ' + str(cur)
            st+= "]]"
            return st
        pass

    def pop(self):
        if self.length==0:
            raise IndexError("LinkedList is empty!")
        elif self.length==1:
            self.head = None
            self.length -=1
        else:
            cur = self.head
            for i in range(self.length-2):
                cur = cur.next
            cur.next = None
            self.length -=1


    def change_on_start(self, n, new_data):
        if n<1:
            raise KeyError("Element doesn't exist!")
        if self.length<n:
            raise KeyError("Element doesn't exist!")
        else:
            cur = self.head
            n = n-1
            for i in range(n):
                cur = cur.next
            cur.data = new_data
        pass

    def clear(self):
        self.head = None
        self.length = 0
        pass

File: CS_dataset/spring/cs-2024-3383/Chernysh_Maksim_lb2/src/lab2.c
class Node:
    def __init__(self, data, next=None):
        self.__data = data
        self.next = next

    def get_data(self):
        return self.__data
        
    def change_data(self, new_data):
        self.__data = new_data

    def __str__(self):
        return f"data: {self.__data}, next: {self.next.__data if self.next != None else None}"


class LinkedList:
    def __init__(self, head = None):
        if head == None:
            self.head = head
            self.__length = 0
        else:
            self.head = head
            self.__length = 1

    def __len__(self):
        return self.__length

    def append(self, element):
        self.__length += 1
        elem = Node(element, None)
        if self.head != None:
            tmp = self.head
            while tmp.next != None:
                tmp = tmp.next
            tmp.next = elem
        else:
            self.head = elem

    def __str__(self):
        if self.head != None:
            tmp = self.head
            elems = f'{tmp}'
            while tmp.next != None:
                tmp = tmp.next
                elems += "; " + tmp.__str__()
            res = f'LinkedList[length = {self.__length}, [{elems}]]'
        else:
            res = f'LinkedList[]'
        return res

    def pop(self):
        if self.head == None:
            raise IndexError("LinkedList is empty!")
        else:
            tmp1 = self.head
            tmp2 = tmp1.next
            while tmp2 != None and tmp2.next != None and tmp2:
                tmp1 = tmp2
                tmp2 = tmp2.next
            if tmp1 == self.head and self.__length == 1:
                self.head = None
            else:
                tmp1.next = None
            self.__length -= 1

    def change_on_start(self, n, new_data):
        if self.__length < n or n<=0:
            raise KeyError("Element doesn't exist!")
        else:
            tmp = self.head
            for i in range(n-1):
                tmp = tmp.next
            tmp.change_data(new_data)

    def clear(self):
        self.head = None
        self.__length = 0
